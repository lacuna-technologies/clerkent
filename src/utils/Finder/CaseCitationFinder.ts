import Constants from '../Constants'
import type Law from '../../types/Law'

export interface CaseCitationFinderResult {
  jurisdiction: Law.JursidictionCode
  citation: string,
  index: number,
  court? : string,
  type: `case-citation`
}

// const emptyParseResult: ParseResult = {
//   citation: undefined,
//   jurisdiction: undefined,
// }

// const parseQuery = (query: string): ParseResult => {
//   const cleanedQuery = query.trim()
//   return parseSGCase(cleanedQuery) || emptyParseResult
// }

// const inSLW = (query: string) => {
//   const eligibleCourt = Array.isArray(query.match(/SGCA|SGHC/))

//   const yearRegex = /^\[(2\d{3})]/
//   const yearMatch = query.match(yearRegex)
//   const eligibleYear = Array.isArray(yearMatch) && Number.parseInt(yearMatch[1]) >= 2000

//   return eligibleCourt && eligibleYear
// }

// const inSGSC = (query: string) => inSLW(query)

const formatAbbrs = (abbrArray) => abbrArray.map(({ abbr, appendum }) => `${abbr
    .split(``)
    .map((letter: string) =>
      /[a-z]/i.test(letter)
        ? letter+`\\.?`
        : letter,
    ).join(``)
    }${appendum ? appendum : ``}`,
  ).join(`|`)

const findSGCaseCitation = (query: string): CaseCitationFinderResult[] => {
  const regex = /\[[12]\d{3}]( \d{1,2})? (sgca(\(i\))?|sghc|sgdc|sgmc|slr(\(r\))?) \d{1,4}/gi
  const matches = [...query.matchAll(regex)]
  if (matches.length > 0) {
    return matches.map((match) => ({
      citation: match[0],
      index: match.index,
      jurisdiction: Constants.JURISDICTIONS.SG.id,
    })).map(c => ({ ...c, type: `case-citation` }))
  }
  return []
}

const findUKCaseCitation = (query:string): CaseCitationFinderResult[] => {
  const abbrs = formatAbbrs([
    { abbr: `EWCA`, appendum: `( Civ| Crim)?` },
    { abbr: `EWHC`, appendum: `( Patents)?` },
    { abbr: `UKSC` },
    { abbr: `UKPC` },
    { abbr: `UKHL` },
    { abbr: `AC` },
    { abbr: `Ch`, appendum: `( D)?` },
    { abbr: `QB`, appendum: `(D)?` },
    { abbr: `KB` },
    { abbr: `WLR`, appendum: `( ?\\(D\\))?` },
    { abbr: `All ER`, appendum: `( \\((D|Comm.?)\\))?` },
    { abbr: `BCLC` },
    { abbr: `BCC` },
    { abbr: `HL Cas` },
    { abbr: `App Cas` },
    { abbr: `Ld Raym` },
    { abbr: `FSR` },
    { abbr: `ECC` },
    { abbr: `ITCLR` },
    { abbr: `RPC` },
    { abbr: `Ex Rep` },
    { abbr: `ER` },
    { abbr: `Cr App R` },
    { abbr: `ALR` },
    { abbr: `FLR` },
    { abbr: `Hare` },
    { abbr: `H & Tw` },
    { abbr: `EMLR` },
    { abbr: `Fam` },
    { abbr: `Macq` },
    { abbr: `TLR` },
    { abbr: `Ves & B` },
    { abbr: `EngR` },
    { abbr: `Lloyd's Rep` },
    { abbr: `BLR` },
    { abbr: `CLC` },
  ])
  // eslint-disable-next-line unicorn/better-regex
  const yearRegex = new RegExp(/((\[|\()[12]\d{3}(-[12]\d{3})?(\]|\)))/)
  const volumeRegex = new RegExp(/( \d{1,2})?/)
  const pageRegex = new RegExp(/\d{1,4}/)
  const regex = new RegExp(`${yearRegex.source}${volumeRegex.source} (${abbrs}) ${pageRegex.source}`, `gi`)

  const matches = [...query.matchAll(regex)]
  if (matches.length > 0) {
    return matches.map((match) => ({
      citation: match[0],
      index: match.index,
      jurisdiction: Constants.JURISDICTIONS.UK.id,
    })).map(c => ({ ...c, type: `case-citation` }))
  }
  return []
}

export const epoRegex = new RegExp(/[GJT][ _]?\d{1,4}\/\d{1,2}/)
export const cjeuRegex = new RegExp(/[CT]-\d{1,3}\/\d{1,2}/)

const findEUCaseCitation = (query: string): CaseCitationFinderResult[] => {
  const regex = new RegExp(`(${epoRegex.source})|(${cjeuRegex.source})`, `gi`)
  const cleanedQuery = query.replace(/â€‘/g, `-`)

  const matches = [...cleanedQuery.matchAll(regex)]
  if(matches.length > 0){
    return matches.map((match) => ({
      citation: match[0],
      court: match[1] ? Constants.COURTS.EU_epo.id : Constants.COURTS.EU_cjeu.id,
      index: match.index,
      jurisdiction: Constants.JURISDICTIONS.EU.id,
    })).map(c => ({ ...c, type: `case-citation` }))
  }
  return []
}

const findHKCaseCitation = (query: string): CaseCitationFinderResult[] => {
  // eslint-disable-next-line unicorn/better-regex
  const yearRegex = new RegExp(/(([([])[12]\d{3}(-[12]\d{3})?[)\]])/)
  const volumeRegex = new RegExp(/( \d{1,2})?/)
  const pageRegex = new RegExp(/\d{1,4}/)
  const abbrs = formatAbbrs([
    { abbr: `HKCA` },
    { abbr: `HKCFA` },
    { abbr: `HKCFI` },
    { abbr: `HKLRD` },
    { abbr: `HKCFAR` },
    { abbr: `HKC` },
    { abbr: `HKDC` },
    { abbr: `HKFC` },
    { abbr: `HKLdT` },
    { abbr: `HKCT` },
    { abbr: `HKCrC` },
    { abbr: `HKOAT` },
    { abbr: `HKLaT` },
    { abbr: `HKMagC` },
  ])
  const regex = new RegExp(`${yearRegex.source}${volumeRegex.source} (${abbrs}) ${pageRegex.source}`, `gi`)

  const matches = [...query.matchAll(regex)]
  if (matches.length > 0) {
    return matches.map((match) => ({
      citation: match[0],
      index: match.index,
      jurisdiction: Constants.JURISDICTIONS.HK.id,
    })).map(c => ({ ...c, type: `case-citation` }))
  }
  return []
}

const findCACaseCitation = (query: string): CaseCitationFinderResult[] => {
  // eslint-disable-next-line unicorn/better-regex
  const yearRegex = new RegExp(/(([([])[12]\d{3}(-[12]\d{3})?[)\]])/)
  const volumeRegex = new RegExp(/( \d{1,2})?/)
  const pageRegex = new RegExp(/\d{1,4}/)
  const abbrs = formatAbbrs([
    { abbr: `SCR` },
    { abbr: `WWR` },
    { abbr: `EXP` },
    { abbr: `CarswellAlta` },
    { abbr: `SCJ No` },
  ])
  const regex = new RegExp(`${yearRegex.source}${volumeRegex.source} (${abbrs}) ${pageRegex.source}`, `gi`)

  const matches = [...query.matchAll(regex)]
  if (matches.length > 0) {
    return matches.map((match) => ({
      citation: match[0],
      index: match.index,
      jurisdiction: Constants.JURISDICTIONS.CA.id,
    })).map(c => ({ ...c, type: `case-citation` }))
  }
  return []
}

const findAUCaseCitation = (query: string): CaseCitationFinderResult[] => {
  // eslint-disable-next-line unicorn/better-regex
  const yearRegex = new RegExp(/(([([])[12]\d{3}(-[12]\d{3})?[)\]])/)
  const volumeRegex = new RegExp(/( \d{1,2})?/)
  const pageRegex = new RegExp(/\d{1,4}/)
  const abbrs = formatAbbrs([
    { abbr: `HCA` },
    { abbr: `UKPCHCA` },
    { abbr: `FamCA` },
    { abbr: `HCASum` },
    { abbr: `NSWSC` },
    { abbr: `NSWCA` },
    { abbr: `NSWCCA` },
    { abbr: `NSWCIMC` },
    { abbr: `NSWCC` },
    { abbr: `NSWDC` },
    { abbr: `NSWDRGC` },
    { abbr: `NSWIC` },
    { abbr: `NSWKnoxRp` },
    { abbr: `NSWLR` },
    { abbr: `NSWR` },
    { abbr: `SR \\(NSW\\)`},
    { abbr: `CLR` },
    { abbr: `FCA` },
    { abbr: `FCAFC` },
    { abbr: `FCCA` },
    { abbr: `FMCA` },
    { abbr: `FMCAfam` },
    { abbr: `IRCA` },
    { abbr: `FCR` },
    { abbr: `ALR` },
    { abbr: `QCA` },
    { abbr: `QSC` },
    { abbr: `QDC` },
    { abbr: `QChC` },
    { abbr: `QChCM` },
    { abbr: `ICQ` },
    { abbr: `QIC` },
    { abbr: `QLC` },
    { abbr: `VSC` },
    { abbr: `VicSC` },
    { abbr: `VSCA` },
    { abbr: `VicCorC` },
    { abbr: `VCC` },
    { abbr: `VMC` },
    { abbr: `VicRp` },
    { abbr: `VicLawRp` },
    { abbr: `WASC` },
    { abbr: `WASCA` },
    { abbr: `WALawRp` },
    { abbr: `WADC` },
    { abbr: `FWCA` },
    { abbr: `FCWAM` },
    { abbr: `WACIC` },
    { abbr: `WAGAB` },
    { abbr: `WASupC` },
    { abbr: `WAR` },
    { abbr: `TASSC` },
    { abbr: `TASFC` },
    { abbr: `TASCCA` },
    { abbr: `TASMC` },
    { abbr: `TASSupC` },
    { abbr: `TASLawRp` },
    { abbr: `TASStRp` },
    { abbr: `TASRp` },
    { abbr: `TASADT` },
    { abbr: `TASFPT` },
    { abbr: `SASC` },
    { abbr: `SASCA` },
    { abbr: `SASCFC` },
    { abbr: `SADC` },
    { abbr: `SAERDC` },
    { abbr: `SALC` },
    { abbr: `SAPelhamRp` },
    { abbr: `SAIRC` },
    { abbr: `SAIndRp` },
    { abbr: `SALawRp` },
    { abbr: `SAStRp` },
    { abbr: `SAWC` },
    { abbr: `SACAT` },
    { abbr: `NTSC` },
    { abbr: `NTCA` },
    { abbr: `NTCCA` },
    { abbr: `NTMC` },
    { abbr: `NTJud` },
    { abbr: `NTLC` },
    { abbr: `ACTCA` },
    { abbr: `ACTSCFC` },
    { abbr: `ACTSC` },
    { abbr: `ACTCD` },
    { abbr: `ACTIC` },
    { abbr: `ACTMC` },
    { abbr: `ACAT` },
    { abbr: `ADO` },
    { abbr: `AUDND` },
    { abbr: `NFSC` },
  ])
  const regex = new RegExp(`${yearRegex.source}${volumeRegex.source} (${abbrs}) ${pageRegex.source}`, `gi`)

  const matches = [...query.matchAll(regex)]
  if (matches.length > 0) {
    return matches.map((match) => ({
      citation: match[0],
      index: match.index,
      jurisdiction: Constants.JURISDICTIONS.AU.id,
    })).map(c => ({ ...c, type: `case-citation` }))
  }
  return []
}

const findNZCaseCitation = (query: string): CaseCitationFinderResult[] => {
  // eslint-disable-next-line unicorn/better-regex
  const yearRegex = new RegExp(/(([([])[12]\d{3}(-[12]\d{3})?[)\]])/)
  const volumeRegex = new RegExp(/( \d{1,2})?/)
  const pageRegex = new RegExp(/\d{1,4}/)
  const abbrs = formatAbbrs([
    { abbr: `NZCA` },
    { abbr: `NZSC` },
    { abbr: `NZHC` },
    { abbr: `NZLR` },
    { abbr: `NZAR` },
    { abbr: `NZFC` },
    { abbr: `NZPC` },
    { abbr: `NZDC` },
  ])
  const regex = new RegExp(`${yearRegex.source}${volumeRegex.source} (${abbrs}) ${pageRegex.source}`, `gi`)

  const matches = [...query.matchAll(regex)]
  if (matches.length > 0) {
    return matches.map((match) => ({
      citation: match[0],
      index: match.index,
      jurisdiction: Constants.JURISDICTIONS.NZ.id,
    })).map(c => ({ ...c, type: `case-citation` }))
  }
  return []
}
const findCaseCitation = (query: string): CaseCitationFinderResult[] => {
  return [
    ...findSGCaseCitation(query),
    ...findUKCaseCitation(query),
    ...findEUCaseCitation(query),
    ...findHKCaseCitation(query),
    ...findCACaseCitation(query),
    ...findAUCaseCitation(query),
    ...findNZCaseCitation(query),
  ]
}

const CaseFinder = {
  findAUCaseCitation,
  findCACaseCitation,
  findCaseCitation,
  findEUCaseCitation,
  findHKCaseCitation,
  findSGCaseCitation,
  findUKCaseCitation,
}

export default CaseFinder