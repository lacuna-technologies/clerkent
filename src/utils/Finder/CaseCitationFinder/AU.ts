import Constants from '../../Constants'
import { formatAbbrs, sortCasesByVolume } from './utils'

export const AUAbbrs = [
  // approximate order only
  // TODO: put this in a proper, definitive order
  { abbr: `HCA` },
  { abbr: `UKPCHCA` },
  { abbr: `FCA` },
  { abbr: `FCAFC` },
  { abbr: `FCCA` },
  { abbr: `FamCA` },
  { abbr: `HCASum` },

  { abbr: `NSWCA` },
  { abbr: `VSCA` },
  { abbr: `QCA` },
  { abbr: `NSWCCA` },
  { abbr: `WASCA` },
  { abbr: `FMCA` },
  { abbr: `FMCAfam` },
  { abbr: `TASCCA` },
  { abbr: `IRCA` },
  { abbr: `FWCA` },
  { abbr: `NTCA` },
  { abbr: `NTCCA` },
  { abbr: `ACTCA` },

  { abbr: `NSWSC` },
  { abbr: `QSC` },
  { abbr: `QSCPR` },
  { abbr: `QSCFC` },
  { abbr: `VicSC` },
  { abbr: `VSC` },
  { abbr: `SASC` },
  { abbr: `WASC` },
  { abbr: `WASupC` },
  { abbr: `TASSC` },
  { abbr: `TASSupC` },
  { abbr: `NTSC` },
  { abbr: `ACTSC` },
  { abbr: `NFSC` },

  { abbr: `NSWLR` },
  { abbr: `NSWR` },
  { abbr: `CLR` },
  { abbr: `WALawRp` },
  { abbr: `SR \\(NSW\\)`},
  { abbr: `FCR` },
  { abbr: `ALR` },
  { abbr: `ALJ` },
  { abbr: `ALJR` },

  { abbr: `NSWCIMC` },
  { abbr: `NSWCC` },
  { abbr: `NSWDC` },
  { abbr: `NSWDRGC` },
  { abbr: `NSWIC` },
  { abbr: `NSWKnoxRp` },
  { abbr: `NSWLC` },
  { abbr: `NSWLEC` },
  { abbr: `NSWIRComm` },
  { abbr: `NSWDRGC` },
  { abbr: `NSWChC` },
  { abbr: `NSWADTAP` },
  { abbr: `NSWADT` },
  { abbr: `NSWCATAP` },
  { abbr: `NSWCATAD` },
  { abbr: `NSWCATCD` },
  { abbr: `NSWCATGD` },
  { abbr: `NSWCATOD ` },
  { abbr: `NSWDDT` },
  { abbr: `NSWEOT` },
  { abbr: `NSWFTT` },
  { abbr: `NSWLST` },
  { abbr: `NSWMT` },
  { abbr: `NSWTAB` },
  { abbr: `QDC` },
  { abbr: `QDCPR` },
  { abbr: `QMC` },
  { abbr: `QCATA` },
  { abbr: `QCAT` },
  { abbr: `QPEC` },
  { abbr: `QLAC` },
  { abbr: `QLC` },
  { abbr: `ICQ` },
  { abbr: `QIRC` },
  { abbr: `QChC` },
  { abbr: `QChCM` },
  { abbr: `QMHC` },
  { abbr: `QIC` },
  { abbr: `QLC` },
  { abbr: `QHPT` },
  { abbr: `VR` },
  { abbr: `VicCorC` },
  { abbr: `VCC` },
  { abbr: `VMC` },
  { abbr: `VicRp` },
  { abbr: `VicLawRp` },
  { abbr: `VLR` },
  { abbr: `TASFC` },
  { abbr: `WADC` },
  { abbr: `FCWAM` },
  { abbr: `WACIC` },
  { abbr: `WAGAB` },
  { abbr: `WAR` },
  { abbr: `SASCA` },
  { abbr: `TASMC` },
  { abbr: `TASLawRp` },
  { abbr: `TASStRp` },
  { abbr: `TASRp` },
  { abbr: `TASADT` },
  { abbr: `TASFPT` },
  { abbr: `Tas SR` },
  { abbr: `SASCFC` },
  { abbr: `SASR` },
  { abbr: `SADC` },
  { abbr: `SAERDC` },
  { abbr: `SALC` },
  { abbr: `SAPelhamRp` },
  { abbr: `SAIRC` },
  { abbr: `SAIndRp` },
  { abbr: `SALawRp` },
  { abbr: `SAStRp` },
  { abbr: `SAWC` },
  { abbr: `SACAT` },
  { abbr: `NTMC` },
  { abbr: `NTJud` },
  { abbr: `NTLC` },
  { abbr: `ACTSCFC` },
  { abbr: `ACTCD` },
  { abbr: `ACTIC` },
  { abbr: `ACTMC` },
  { abbr: `ACAT` },
  { abbr: `ADO` },
  { abbr: `AUDND` },
  { abbr: `ACSR` },
  { abbr: `ACLC` },
  { abbr: `ATC` }, // Australian Tax Cases
  { abbr: `ACLR` },
]

export const sortAUCases = (
  citationsArray: Law.Case[],
  attribute: string,
): Law.Case[] => sortCasesByVolume(AUAbbrs, citationsArray, attribute)

export const findAUCaseCitationMatches = (query: string) => {
   // eslint-disable-next-line unicorn/better-regex
  const yearRegex = new RegExp(/(([([])[12]\d{3}(-[12]\d{3})?[)\]])/)
  const volumeRegex = new RegExp(/( \d{1,3})?/)
  const pageRegex = new RegExp(/\d{1,4}/)
  const abbrs = formatAbbrs(AUAbbrs)
  const regex = new RegExp(
    `(?<year>${yearRegex.source})`+
    `(?<volume>${volumeRegex.source}) `+
    `(?<abbr>${abbrs}) `+
    `${pageRegex.source}`, `gi`)

  return [...query.matchAll(regex)]
}

export const findAUCaseCitation = (query: string): Finder.CaseCitationFinderResult[] => {
  const matches = findAUCaseCitationMatches(query)
  if (matches.length > 0) {
    return matches.map((match) => ({
      abbr: match.groups.abbr,
      citation: match[0],
      index: match.index,
      jurisdiction: Constants.JURISDICTIONS.AU.id,
      year: match.groups.year,
    })).map(c => ({ ...c, type: `case-citation` }))
  }
  return []
}