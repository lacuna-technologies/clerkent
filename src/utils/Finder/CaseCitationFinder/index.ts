export { default } from './CaseCitationFinder'
export type { CaseCitationFinderResult } from './types'