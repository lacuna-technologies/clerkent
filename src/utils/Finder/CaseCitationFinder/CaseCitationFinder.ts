import { findUKCaseCitation, sortUKCitations } from './UK'
import { findSGCaseCitation , sortSGCitations } from './SG'
import { findNZCaseCitation, sortNZCitations } from './NZ'
import { findHKCaseCitation, sortHKCitations } from './HK'
import { findCACaseCitation, sortCACitations } from './CA'
import { findAUCaseCitation, sortAUCitations } from './AU'
import { findEUCaseCitation, sortEUCitations } from './EU'
import { findMYCaseCitation, sortMYCitations } from './MY'
import { findECHRCaseCitation, sortECHRCitations } from './ECHR'
import { findUNCaseCitation, sortUNCitations } from './UN'

const findCaseCitation = (query: string): Finder.CaseCitationFinderResult[] => {
  return [
    ...findSGCaseCitation(query),
    ...findUKCaseCitation(query),
    ...findEUCaseCitation(query),
    ...findHKCaseCitation(query),
    ...findCACaseCitation(query),
    ...findAUCaseCitation(query),
    ...findNZCaseCitation(query),
    ...findMYCaseCitation(query),
    ...findECHRCaseCitation(query),
    ...findUNCaseCitation(query),
  ]
}

const CaseFinder = {
  findAUCaseCitation,
  findCACaseCitation,
  findCaseCitation,
  findECHRCaseCitation,
  findEUCaseCitation,
  findHKCaseCitation,
  findNZCaseCitation,
  findSGCaseCitation,
  findUKCaseCitation,
  findUNCaseCitation,
  sortAUCitations,
  sortCACitations,
  sortECHRCitations,
  sortEUCitations,
  sortHKCitations,
  sortMYCitations,
  sortNZCitations,
  sortSGCitations,
  sortUKCitations,
  sortUNCitations,
}

export default CaseFinder