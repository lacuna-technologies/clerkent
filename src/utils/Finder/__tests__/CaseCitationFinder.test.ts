import CaseCitationFinder from '../CaseCitationFinder'

describe(`Finder`, () => {
  it(`ðŸ‡¸ðŸ‡¬ citations`, () => {
    const citations = [
      `[2021] SGCA 34`,
      `[2020] SGHC 228`,
      `[2020] SGHC 100 (2010)`,
      `[2011] 2 SLR 47`,
      `[2020] SGDC 95`,
      `[1994] SGCA 128`,
      `[1988] SGCA 16`,
      `[2001] 2 SLR 253`,
      `[2001] 2 SLR 253`,
    ]

    for(const citation of citations){
      const parsedCitation = CaseCitationFinder.findCaseCitation(citation)
      expect(parsedCitation.length).toEqual(1)
      expect(parsedCitation[0].jurisdiction).toEqual(`SG`)
      expect(parsedCitation).toMatchSnapshot()
    }
  })

  it(`ðŸ‡¬ðŸ‡§ citations`, () => {
    const citations = [
      `(1754) 2 Ves Sen 547`,
      `[1843-60] All ER 249`,
      `[1948] 1 KB 223`,
      `[1965] 1 QB 456`,
      `(1951) 35 Cr App R 164`,
      `[2007] UKHL 17`,
      `O/328/22`,
      `O/002/22`,
    ]

    for (const citation of citations) {
      const parsedCitation = CaseCitationFinder.findCaseCitation(citation)
      expect(parsedCitation.length).toEqual(1)
      expect(parsedCitation[0].jurisdiction).toEqual(`UK`)
      expect(parsedCitation).toMatchSnapshot()
    }
  })
})