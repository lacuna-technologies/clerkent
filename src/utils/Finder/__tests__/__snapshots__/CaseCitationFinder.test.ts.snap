// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
Array [
  Object {
    "abbr": "SGCA",
    "citation": "[2021] SGCA 34",
    "index": 0,
    "jurisdiction": "SG",
    "type": "case-citation",
    "year": "2021",
  },
]
`;

exports[` 2`] = `
Array [
  Object {
    "abbr": "SGHC",
    "citation": "[2020] SGHC 228",
    "index": 0,
    "jurisdiction": "SG",
    "type": "case-citation",
    "year": "2020",
  },
]
`;

exports[` 3`] = `
Array [
  Object {
    "abbr": "SGHC",
    "citation": "[2020] SGHC 100",
    "index": 0,
    "jurisdiction": "SG",
    "type": "case-citation",
    "year": "2020",
  },
]
`;

exports[` 4`] = `
Array [
  Object {
    "abbr": "SLR",
    "citation": "[2011] 2 SLR 47",
    "index": 0,
    "jurisdiction": "SG",
    "type": "case-citation",
    "year": "2011",
  },
]
`;

exports[` 5`] = `
Array [
  Object {
    "abbr": "SGDC",
    "citation": "[2020] SGDC 95",
    "index": 0,
    "jurisdiction": "SG",
    "type": "case-citation",
    "year": "2020",
  },
]
`;

exports[` 6`] = `
Array [
  Object {
    "abbr": "SGCA",
    "citation": "[1994] SGCA 128",
    "index": 0,
    "jurisdiction": "SG",
    "type": "case-citation",
    "year": "1994",
  },
]
`;

exports[` 7`] = `
Array [
  Object {
    "abbr": "SGCA",
    "citation": "[1988] SGCA 16",
    "index": 0,
    "jurisdiction": "SG",
    "type": "case-citation",
    "year": "1988",
  },
]
`;

exports[` 8`] = `
Array [
  Object {
    "abbr": "SLR",
    "citation": "[2001] 2 SLR 253",
    "index": 0,
    "jurisdiction": "SG",
    "type": "case-citation",
    "year": "2001",
  },
]
`;

exports[` 9`] = `
Array [
  Object {
    "abbr": "SLR",
    "citation": "[2001] 2 SLR 253",
    "index": 0,
    "jurisdiction": "SG",
    "type": "case-citation",
    "year": "2001",
  },
]
`;

exports[` 10`] = `
Array [
  Object {
    "abbr": "SGIPOS",
    "citation": "[2022] SGIPOS 4",
    "index": 0,
    "jurisdiction": "SG",
    "type": "case-citation",
    "year": "2022",
  },
]
`;

exports[` 11`] = `
Array [
  Object {
    "abbr": "SGHCR",
    "citation": "[2022] SGHCR 1",
    "index": 0,
    "jurisdiction": "SG",
    "type": "case-citation",
    "year": "2022",
  },
]
`;

exports[` 12`] = `
Array [
  Object {
    "abbr": "SGHCF",
    "citation": "[2020] SGHCF 18",
    "index": 0,
    "jurisdiction": "SG",
    "type": "case-citation",
    "year": "2020",
  },
]
`;

exports[` 13`] = `
Array [
  Object {
    "abbr": "SGHC(A)",
    "citation": "[2022] SGHC(A) 17",
    "index": 0,
    "jurisdiction": "SG",
    "type": "case-citation",
    "year": "2022",
  },
]
`;

exports[` 14`] = `
Array [
  Object {
    "abbr": "SGCA(I)",
    "citation": "[2020] SGCA(I) 2",
    "index": 0,
    "jurisdiction": "SG",
    "type": "case-citation",
    "year": "2020",
  },
]
`;

exports[` 15`] = `
Array [
  Object {
    "abbr": "Ves Sen",
    "citation": "(1754) 2 Ves Sen 547",
    "index": 0,
    "jurisdiction": "UK",
    "type": "case-citation",
    "year": "1754",
  },
]
`;

exports[` 16`] = `
Array [
  Object {
    "abbr": "All ER",
    "citation": "[1843-60] All ER 249",
    "index": 0,
    "jurisdiction": "UK",
    "type": "case-citation",
    "year": "1843",
  },
]
`;

exports[` 17`] = `
Array [
  Object {
    "abbr": "KB",
    "citation": "[1948] 1 KB 223",
    "index": 0,
    "jurisdiction": "UK",
    "type": "case-citation",
    "year": "1948",
  },
]
`;

exports[` 18`] = `
Array [
  Object {
    "abbr": "QB",
    "citation": "[1965] 1 QB 456",
    "index": 0,
    "jurisdiction": "UK",
    "type": "case-citation",
    "year": "1965",
  },
]
`;

exports[` 19`] = `
Array [
  Object {
    "abbr": "Cr App R",
    "citation": "(1951) 35 Cr App R 164",
    "index": 0,
    "jurisdiction": "UK",
    "type": "case-citation",
    "year": "1951",
  },
]
`;

exports[` 20`] = `
Array [
  Object {
    "abbr": "UKHL",
    "citation": "[2007] UKHL 17",
    "index": 0,
    "jurisdiction": "UK",
    "type": "case-citation",
    "year": "2007",
  },
]
`;

exports[` 21`] = `
Array [
  Object {
    "abbr": "UKIPO",
    "citation": "O/328/22",
    "index": 0,
    "jurisdiction": "UK",
    "type": "case-citation",
    "year": "2022",
  },
]
`;

exports[` 22`] = `
Array [
  Object {
    "abbr": "UKIPO",
    "citation": "O/002/22",
    "index": 0,
    "jurisdiction": "UK",
    "type": "case-citation",
    "year": "2022",
  },
]
`;

exports[` 23`] = `
Array [
  Object {
    "abbr": "UKIPO",
    "citation": "O01422",
    "index": 0,
    "jurisdiction": "UK",
    "type": "case-citation",
    "year": "2022",
  },
]
`;

exports[` 24`] = `
Array [
  Object {
    "abbr": "EWHC",
    "citation": "[2021] EWHC 2168",
    "index": 0,
    "jurisdiction": "UK",
    "type": "case-citation",
    "year": "2021",
  },
]
`;

exports[` 25`] = `
Array [
  Object {
    "abbr": "EWHC",
    "citation": "[2016] EWHC 3151",
    "index": 0,
    "jurisdiction": "UK",
    "type": "case-citation",
    "year": "2016",
  },
]
`;

exports[` 26`] = `
Array [
  Object {
    "abbr": "ECDR",
    "citation": "[2001] ECDR 5",
    "index": 0,
    "jurisdiction": "UK",
    "type": "case-citation",
    "year": "2001",
  },
]
`;

exports[` 27`] = `
Array [
  Object {
    "citation": "C-40/17",
    "court": "CJEU",
    "index": 0,
    "jurisdiction": "EU",
    "type": "case-citation",
  },
]
`;

exports[` 28`] = `
Array [
  Object {
    "citation": "C-403/08",
    "court": "CJEU",
    "index": 6,
    "jurisdiction": "EU",
    "type": "case-citation",
  },
  Object {
    "citation": "C-429/08",
    "court": "CJEU",
    "index": 19,
    "jurisdiction": "EU",
    "type": "case-citation",
  },
]
`;

exports[` 29`] = `
Array [
  Object {
    "citation": "C-145/10",
    "court": "CJEU",
    "index": 2,
    "jurisdiction": "EU",
    "type": "case-citation",
  },
]
`;

exports[` 30`] = `
Array [
  Object {
    "citation": "C-5/08",
    "court": "CJEU",
    "index": 2,
    "jurisdiction": "EU",
    "type": "case-citation",
  },
]
`;

exports[` 31`] = `
Array [
  Object {
    "abbr": "FCA",
    "citation": "[2010] FCA 984",
    "index": 0,
    "jurisdiction": "AU",
    "type": "case-citation",
    "year": "[2010]",
  },
]
`;

exports[` 32`] = `
Array [
  Object {
    "abbr": "CLR",
    "citation": "(1990) 171 CLR 30",
    "index": 0,
    "jurisdiction": "AU",
    "type": "case-citation",
    "year": "(1990)",
  },
]
`;

exports[` 33`] = `
Array [
  Object {
    "abbr": "ACSR",
    "citation": "(1992) 7 ACSR 759",
    "index": 0,
    "jurisdiction": "AU",
    "type": "case-citation",
    "year": "(1992)",
  },
]
`;

exports[` 34`] = `
Array [
  Object {
    "abbr": "ALJ",
    "citation": "(2003) 200 ALJ 1",
    "index": 0,
    "jurisdiction": "AU",
    "type": "case-citation",
    "year": "(2003)",
  },
]
`;

exports[` 35`] = `
Array [
  Object {
    "abbr": "HCA",
    "citation": "[1989] HCA 9",
    "index": 0,
    "jurisdiction": "AU",
    "type": "case-citation",
    "year": "[1989]",
  },
]
`;

exports[` 36`] = `
Array [
  Object {
    "abbr": "ACLR",
    "citation": "(1982) 7 ACLR 171",
    "index": 0,
    "jurisdiction": "AU",
    "type": "case-citation",
    "year": "(1982)",
  },
]
`;

exports[` 37`] = `
Array [
  Object {
    "citation": "[1999] 2 SCR 534",
    "index": 0,
    "jurisdiction": "CA",
    "type": "case-citation",
  },
]
`;

exports[` 38`] = `
Array [
  Object {
    "citation": "(2008) 91 OR (3d) 353",
    "index": 0,
    "jurisdiction": "CA",
    "type": "case-citation",
  },
]
`;

exports[` 39`] = `
Array [
  Object {
    "citation": "(1973) 40 DLR (3d) 371",
    "index": 0,
    "jurisdiction": "CA",
    "type": "case-citation",
  },
]
`;

exports[` 40`] = `
Array [
  Object {
    "citation": "[1983] 4 WWR 762",
    "index": 0,
    "jurisdiction": "CA",
    "type": "case-citation",
  },
]
`;

exports[`Finder [2021] SGCA 34 1`] = `
Array [
  Object {
    "citation": "[2006] SCC 39",
    "index": 0,
    "jurisdiction": "CA",
    "type": "case-citation",
  },
]
`;

exports[`Finder [2021] SGCA 34 2`] = `
Array [
  Object {
    "citation": "(2013) 16 HKCFAR 366",
    "index": 0,
    "jurisdiction": "HK",
    "type": "case-citation",
  },
]
`;

exports[`Finder [2021] SGCA 34 3`] = `
Array [
  Object {
    "citation": "[2015] HKCFI 1856",
    "index": 0,
    "jurisdiction": "HK",
    "type": "case-citation",
  },
]
`;

exports[`Finder [2021] SGCA 34 4`] = `
Array [
  Object {
    "citation": "[1999] HKCA 585",
    "index": 0,
    "jurisdiction": "HK",
    "type": "case-citation",
  },
]
`;

exports[`Finder [2021] SGCA 34 5`] = `
Array [
  Object {
    "citation": "[2000] 1 HKLRD 595",
    "index": 0,
    "jurisdiction": "HK",
    "type": "case-citation",
  },
]
`;

exports[`Finder [2021] SGCA 34 6`] = `
Array [
  Object {
    "citation": "[2012] HKDC 911",
    "index": 0,
    "jurisdiction": "HK",
    "type": "case-citation",
  },
]
`;
